src/Data/Text.hs

```diff
@@ -1522,17 +1525,29 @@ group = groupBy (==)
 -- | /O(n)/ Return all initial segments of the given 'Text', shortest
 -- first.
 inits :: Text -> [Text]
-inits t = empty : case t of
+inits = NonEmptyList.toList . initsNE
+
+-- | /O(n)/ Return all initial segments of the given 'Text', shortest
+-- first.
+initsNE :: Text -> NonEmptyList.NonEmpty Text
+initsNE t = empty NonEmptyList.:| case t of
   Text arr off len ->
-    let loop i | i >= len = []
-               | otherwise = let !j = i + iter_ t i in Text arr off j : loop j
+    let loop i
+          | i >= len = []
+          | otherwise = let !j = i + iter_ t i in Text arr off j : loop j
     in loop 0
 
 -- | /O(n)/ Return all final segments of the given 'Text', longest
 -- first.
 tails :: Text -> [Text]
-tails t | null t    = [empty]
-        | otherwise = t : tails (unsafeTail t)
+tails = NonEmptyList.toList . tailsNE
```

Comment:
IIRC it's important to use 
```
notThatLazyToList (a :| as) -> a : as
```
instead of [`NonEmptyList.toList`](https://hackage.haskell.org/package/base-4.19.0.0/docs/src/Data.List.NonEmpty.html#toList) (watch for `~`) so that stuff gets inlined and optimized away.

Reply:
Strictness was added by writing `(toList $!)` instead of `toList`. Hopefully this will work.